easybuild/tools/toolchain/compiler.py

```diff
@@ -278,18 +279,40 @@ def _set_optimal_architecture(self, default_optarch=None):
                                 (--optarch and --optarch=GENERIC still override this value)
         """
         optarch = None
+        use_generic = False
         # --optarch is specified with flags to use
-        if build_option('optarch') is not None and build_option('optarch') != OPTARCH_GENERIC:
-            optarch = build_option('optarch')
-        # --optarch=GENERIC
-        elif build_option('optarch') == OPTARCH_GENERIC:
+        if build_option('optarch') is not None:
+            for optarch_option in build_option('optarch').split(','):
+                # No compilers specified, and option is not generic, so simply take what is there
+                if ":" not in optarch_option and optarch_option != OPTARCH_GENERIC:
+                    optarch = optarch_option
+                    break
+                # Option is generic, so mark it as to be set later and stop iterating
+                elif optarch_option == OPTARCH_GENERIC:
+                    use_generic = True
+                    break
+                # Compilers specified, check if it is the one we need
+                else:
+                    current_compiler = getattr(self, 'COMPILER_FAMILY', None)
+                    comp = optarch_option.split(":")[0]
+                    comp_optarch = optarch_option.split(":")[1] if len(optarch_option.split(":")) > 1 else None 
+                    if comp == current_compiler:
```

Comment:
you're not both parsing the `--optarch` value and checking for a match in one go, it would make more sense to do the parsing a lot earlier (when initialising the toolchain), to catch any formatting problems with the specified `--optarch` value early on

the actual checking of which flags to use from that parsed value should still occur here